calico/felix/frules.py

```diff
@@ -16,7 +16,93 @@
+* the FORWARD chain if they are to be forwarded between two interfaces.
+
+Note: packets that are being forwarded do not traverse the INPUT or
+OUTPUT chains at all.  INPUT and OUTPUT are only used for packets that
+the host itself is to receive/send.
+
+Packet paths
+~~~~~~~~~~~~
+
+There are a number of possible paths through the filter chains that we
+care about:
+
+* Packets from a local workload to another local workload traverse the
+  FORWARD chain only.  Felix must ensure that those packets have *both*
+  the outbound policy of the sending workload and the inbound policy
+  of the receiving workload applied.
+
+* Packets from a local workload to a remote address traverse the FORWARD
+  chain only.  Felix must ensure that those packets have the outbound
+  policy of the local workload applied.
+
+* Packets from a remote address to a local workload traverse the FORWARD
+  chain only.  Felix must apply the inbound policy of the local workload.
+
+* Packets from a local workload to the host itself traverse the INPUT
+  chain.  Felix must apply the outbound policy of the workload.
+
+Chain structure
+~~~~~~~~~~~~~~~
+
+Rather than adding many rules to the kernel chains, which are a shared
+resource (and hence difficult to unpick), Felix creates its own delegate
+chain for each kernel chain and inserts a single jump rule into the
+kernel chain:
+
+* INPUT -> felix-INPUT
+* FORWARD -> felix-FORWARD
+
+The top-level felix-XXX chains are static and configured at start-of-day.
+
+The felix-FORWARD chain sends packet that arrive from a local workload to
+the felix-FROM-ENDPOINT chain, which applies inbound policy.  Packets that
+are denied by policy are dropped immediately.  However, accepted packets
+are returned to the felix-FORWARD chain in case they need to be processed
+further.  felix-FORWARD then directs packets that are going to local
+endpoints to the felix-TO-ENDPOINT chain, which applies inbound policy.
+Similarly, felix-TO-ENDPOINT either drops or returns the packet.  Finally,
+if both the FROM-ENDPOINT and TO-ENDPOINT chains allow the packet,
+felix-FORWARD accepts the packet and allows it through.
+
+The felix-INPUT sends packets from local workloads to the (shared)
+felix-FROM-ENDPOINT chain, which applies outbound policy.  Then it
+(optionally) accepts packets that are returned.
+
+Since workloads come and go, the TO/FROM-ENDPOINT chains are dynamic and
+consist of dispatch tables based on device name.  Those chains are managed
+by dispatch.py.
+
+The dispatch chains direct packets to per-endpoint ("felix-to/from")
+chains, which are responsible for policing IP addresses.  Those chains are
+managed by endpoint.py.  Since the actual policy rules can be shared by
+multiple endpoints, we put each set of policy rules in its own chain and
+the per-endpoint chains send packets to the relevant policy
+(felix-p-xxx-i/o) chains in turn.  Policy profile chains are managed by
+profilerules.py.
+
```

Comment:
This is all great, but I still wasn't really clearly seeing the big picture, and in particular it wasn't obvious to me why a profile chain has 3 possible return values, rather than just 2 (deny or accept)?

I've worked it all out now, and the main problem was that I didn't have enough of the etcd data model in my head, and needed to revise that.  But, perhaps we could add even more hear to make that immediately clearer to other readers as well.

Here are my notes from working this out:

    Endpoint outbound security
     Profile 1
     Profile 2
     ...
     DROP if not matched by any profiles
    
    
    
    Endpoint inbound security
     Profile 1
     Profile 2
     ...
     DROP if not matched by any profiles
    
    
    Profile:
     Tags
     Rules
      Allow/deny stuff to/from IP address/prefix/tag
    
    
    Matched, Deny -> DROP
    Matched, Allow -> RETURN with mark 1
    No match -> RETURN with no mark (or mark 0)
    
    
    "Default behaviour if no rules are matched or if there are no rules in
    the list or it is missing is to deny."

I think the key points are just that

- an individual profile doesn't have an overall 'deny' or 'accept' default behaviour; it only expresses an opinion on the particular IP addresses, prefixes or tags that it matches

- if a given packet isn't matched by any profiles, there is a global overall default of 'deny'.

I wonder if it would be worth adding an expression of these points, and/or some of my notes, here?